src/qibolab/transpilers/general_connectivity.py

```diff
@@ -377,16 +472,83 @@ def add_swaps(self, path, meeting_point):
+                self._transpiled_circuit.add(gates.SWAP(self._qubit_map[f1], self._qubit_map[f2]))
         if len(backward) > 1:
             for b1, b2 in pairwise(backward):
-                self.transpiled_circuit.add(gates.SWAP(self._qubit_map[b1], self._qubit_map[b2]))
+                self._transpiled_circuit.add(gates.SWAP(self._qubit_map[b1], self._qubit_map[b2]))
 
     def update_qubit_map(self):
         """Update the qubit mapping after adding swaps"""
         old_mapping = self._qubit_map.copy()
         for key, value in self._mapping.items():
             self._qubit_map[value] = old_mapping[key]
+
+
+def translate_circuit(circuit, two_qubit_natives, translate_single_qubit=False):
+    """Translates a circuit to native gates.
+
+    Args:
+        circuit (qibo.models.Circuit): Circuit model to translate into native gates.
+        two_qubit_natives (list): List of two qubit native gates
+            supported by the quantum hardware ("CZ" and/or "iSWAP").
+        translate_single_qubit (bool):
+
+    Returns:
+        new (qibo.models.Circuit): Equivalent circuit with native gates.
+    """
+    new = type(circuit)(circuit.nqubits)
+    for gate in circuit.queue:
+        if len(gate.qubits) > 1 or translate_single_qubit:
+            new.add(translate_gate(gate, two_qubit_natives))
+        else:
+            new.add(gate)
+    return new
+
+
+def can_execute(circuit: Circuit, two_qubit_natives: TwoQubitNatives, connectivity: nx.Graph, verbose=True):
+    """Checks if a circuit can be executed on Hardware.
+
+    Args:
+        circuit (qibo.models.Circuit): Circuit model to check.
+        two_qubit_natives (TwoQubitNatives): two qubit gate/s that can be implemented by the hardware.
+        connectivity (networkx.graph): chip connectivity.
+        verbose (bool): If ``True`` it prints debugging log messages.
+
+    Returns ``True`` if the following conditions are satisfied:
+        - Circuit does not contain more than two-qubit gates.
+        - All one-qubit gates are I, Z, RZ or U3.
+        - All two-qubit gates are CZ or iSWAP based on two_qubit_natives.
+        - Circuit matches connectivity.
+    otherwise returns ``False``.
+    """
+
+    # pring messages only if ``verbose == True``
+    vlog = lambda msg: log.info(msg) if verbose else lambda msg: None
+    for gate in circuit.queue:
+        if isinstance(gate, gates.M):
+            continue
+
+        if len(gate.qubits) == 1:
+            if not isinstance(gate, (gates.I, gates.Z, gates.RZ, gates.U3)):
+                vlog(f"{gate.name} is not a single qubit native gate.")
+                return False
+
+        elif len(gate.qubits) == 2:
+            try:
+                if not (TwoQubitNatives.from_gate(gate) in two_qubit_natives):
+                    vlog(f"{gate.name} is not in two_qubit_native.")
+                    return False
+            except ValueError:
```

Comment:
@AleCandido @stavros11  Here the exception is not working correctly. This can be seen in the test test_cannot_execute_wrong_native(), where can_execute is not returning False (i have tried putting Attribute error but is still not working)

Reply:
I think i have solved this